### ISSUE TITLE
Add more SQL/JSON support for the SNOWFLAKE dialect

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14661]
Continuing after https://github.com/jOOQ/jOOQ/issues/14653 from jOOQ 3.18, we could consider adding more SQL/JSON support to the Snowflake dialect, including:

Snowflake supports numerous SQL/JSON operator equivalents, including:

- [ ] `JSON_OBJECTAGG`: https://docs.snowflake.com/en/sql-reference/functions/object_agg
  - [ ] `NULL ON NULL` isn't supported and I don't see any way to emulate it.
- [ ] `JSON_ARRAYAGG`: https://docs.snowflake.com/en/sql-reference/functions/array_agg
  - [ ] `NULL ON NULL` isn't supported and I don't see any way to emulate it.
- [ ] `JSON_GET_ELEMENT`: https://docs.snowflake.com/en/sql-reference/functions/get
- [ ] `JSON_GET_ATTRIBUTE`: https://docs.snowflake.com/en/sql-reference/functions/get
- [ ] `JSON_REMOVE`: https://docs.snowflake.com/en/sql-reference/functions/object_delete
- [ ] `JSON_INSERT`: https://docs.snowflake.com/en/sql-reference/functions/object_insert
- [ ] `JSON_TABLE`: https://docs.snowflake.com/en/sql-reference/functions/flatten

Further tasks:

- [ ] Parsing `ARRAY_AGG` should become parse dialect specific

**Comment by Piscator-FiMo** [comment]
Hi @lukaseder,
are there plans to support flatten as a SQL/JSON function?

**Comment by lukaseder** [comment]
Thanks for your message, @Piscator-FiMo. It's on the list for `JSON_TABLE` emulation, though I haven't investigated yet, if it's really possible. Perhaps along with some `PIVOT` usage? Did you have any particular jOOQ API in mind, other than standard SQL/JSON `JSON_TABLE`?

**Comment by Piscator-FiMo** [comment]
Thanks for the quick response. 
I didn't have any particular API in mind but JSON_TABLE should work as flatten essentially works like a udtf. 
Support for this is coming if I'm not mistak e.g (#13530)](https://github.com/jOOQ/jOOQ/issues/13530)

**Comment by lukaseder** [comment]
#13530 is about https://www.jooq.org/doc/latest/manual/sql-building/procedural-statements.

What you're looking for is code generation and call support for user defined function, which has been implemented indeed: https://github.com/jOOQ/jOOQ/issues/14648. That doesn't include built-in user defined functions, like `FLATTEN`, though. I can look into `FLATTEN` again, this week, but can't promise it'll work out via `JSON_TABLE` emulations.

Of course, you can always just create a plain SQL template as a workaround:
https://www.jooq.org/doc/latest/manual/sql-building/plain-sql-templating/

**Comment by Piscator-FiMo** [comment]
> Of course, you can always just create a plain SQL template as a workaround: https://www.jooq.org/doc/latest/manual/sql-building/plain-sql-templating/

This is what I'm currently doing, but jooq support would be appreciated ^^, I will look into #14648 to update my implementation. 
Thanks

**Comment by lukaseder** [comment]
Great! Note, 3.18 will ship soon, so if you can test 3.18.0-SNAPSHOT versions and provide feedback, that might still be considered prior to shipping.

**Comment by Piscator-FiMo** [comment]
Hi @lukaseder,
I have tested the SNAPSHOT Version a bit. As preliminary feedback I can say the Code generation with VARIANT and ARRAY Datatypes seems to work. I noticed that VARIANT get converted to JSONB and ARRAY to JSON this could lead to code that technically works but wouldn't produce a result. Also I didn't find ARRAY functions such as ARRAY_CONTAINS, ARRAY_COMPACT in the DSL API etc. (https://docs.snowflake.com/en/sql-reference/functions-semistructured)
Furthermore I noticed that some combinations of valid jooq Code lead to invalid SQL. 
e.g:
`dslContext.select(TABLENAME.FIELD.contains(DSL.jsonArray(DSL.val("someValue"))` 
where FIELD is of type<Record, JSON> and of DB type ARRAY.

I am also still missing the equal_null and flatten functionality from Snowflake. 
Equal_null is important because the Snowflake Dialect does some funky things with the classic equals method.
I hope this helps

**Comment by lukaseder** [comment]
> I noticed that VARIANT get converted to JSONB and ARRAY to JSON this could lead to code that technically works but wouldn't produce a result.

Ah yes, great feedback. That should be fixed right away, it would be inconvenient for users.

> Also I didn't find ARRAY functions such as ARRAY_CONTAINS, ARRAY_COMPACT in the DSL API etc. (https://docs.snowflake.com/en/sql-reference/functions-semistructured)

- `ARRAY_CONTAINS`: should be supported via `Field<JSON>.contains(...)` indeed. But it isn't so simple. Other dialects are much more powerful than Snowflake, it seems. Standardising over these functions isn't trivial. I've created a separate issue to track this: https://github.com/jOOQ/jOOQ/issues/14737
- `ARRAY_COMPACT`: That's just Snowflake's way of say `ABSENT ON NULL` (the SQL standard way of removing `NULL` values from JSON arrays and objects). You can use that clause in jOOQ, e.g. `jsonArray(...).absentOnNull()`

Let me know if you find other functions of interest. I've checked them all, and found that they're all hard to represent directly in jOOQ's API without further investigation.

> Furthermore I noticed that some combinations of valid jooq Code lead to invalid SQL.

Yeah, see above. That should probably render `ARRAY_CONTAINS`, although I don't want to rush this without considering all the dialects, first, in #14737

> I am also still missing the equal_null 

You don't really need it? The manual lists that it is equivalent to the `DISTINCT` predicate, which jOOQ supports out of the box: https://docs.snowflake.com/en/sql-reference/functions/equal_null

Though, we could support it in the parser.

> and flatten functionality from Snowflake.

`FLATTEN` no doubt has its value, though it is very vendor specific. If I can manage to emulate `JSON_TABLE` using `FLATTEN`, then sure. But there needs to be some additional research into what constitutes a "path" in Snowflake. It isn't a standard JSON path, from what I can tell. So, what does this mean for jOOQ? Should we patch the JSON path that you provide in the API (e.g. turn `$.a.b` into `a.b`)? Or should we support both? Or should we leave it untouched?

**Comment by Piscator-FiMo** [comment]
> FLATTEN no doubt has its value, though it is very vendor specific. If I can manage to emulate JSON_TABLE using FLATTEN, then sure. But there needs to be some additional research into what constitutes a "path" in Snowflake. It isn't a standard JSON path, from what I can tell. So, what does this mean for jOOQ? Should we patch the JSON path that you provide in the API (e.g. turn $.a.b into a.b)? Or should we support both? Or should we leave it untouched?

I agree it isn't straigth forward. As a sidenote I find it is easy to implement a simple solution to traverse the JSON Path for Snowflake on top of the jooq jsonGetAttribute function so the issue of JSON path traversal is not really pressing for me at least.
Thanks for the reply I will keep your points in mind.
